File: medium/0018_4sum/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> fourSum(std::vector<int> &nums, int target) {
    const size_t numsLen = nums.size();
    if (numsLen < 4)
      return {};

    std::sort(nums.begin(), nums.end());

    std::vector<std::vector<int>> fourSums;

    for (size_t outerLeft = 0; outerLeft < numsLen - 3; ++outerLeft) {
      if (outerLeft > 0 && nums[outerLeft] == nums[outerLeft - 1])
        continue;

      if ((long)nums[outerLeft] + nums[outerLeft + 1] + nums[outerLeft + 2] +
              nums[outerLeft + 3] >
          target)
        break;
      if ((long)nums[outerLeft] + nums[numsLen - 3] + nums[numsLen - 2] +
              nums[numsLen - 1] <
          target)
        continue;

      for (size_t innerLeft = outerLeft + 1; innerLeft < numsLen - 2;
           ++innerLeft) {
        if (innerLeft > outerLeft + 1 && nums[innerLeft] == nums[innerLeft - 1])
          continue;

        if ((long)nums[outerLeft] + nums[innerLeft] + nums[innerLeft + 1] +
                nums[innerLeft + 2] >
            target)
          break;
        if ((long)nums[outerLeft] + nums[innerLeft] + nums[numsLen - 2] +
                nums[numsLen - 1] <
            target)
          continue;

        size_t innerRight = innerLeft + 1;
        size_t outerRight = numsLen - 1;

        long leftSum = nums[outerLeft] + nums[innerLeft];

        while (innerRight < outerRight) {
          long currSum = leftSum + nums[innerRight] + nums[outerRight];

          if (currSum == target) {
            fourSums.push_back({nums[outerLeft], nums[innerLeft],
                                nums[innerRight], nums[outerRight]});

            while (innerRight < outerRight &&
                   nums[innerRight] == nums[innerRight + 1])
              ++innerRight;
            while (innerRight < outerRight &&
                   nums[outerRight] == nums[outerRight - 1])
              --outerRight;

            ++innerRight;
            --outerRight;
          } else if (currSum < target)
            ++innerRight;
          else
            --outerRight;
        }
      }
    }

    return fourSums;
  }
};

File: medium/0018_4sum/solution.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        nums_len = len(nums)

        four_sums = []

        for outer_left in range(nums_len - 3):
            if outer_left > 0 and nums[outer_left] == nums[outer_left - 1]:
                continue

            if (
                nums[outer_left]
                + nums[outer_left + 1]
                + nums[outer_left + 2]
                + nums[outer_left + 3]
                > target
            ):
                break
            if nums[outer_left] + nums[-3] + nums[-2] + nums[-1] < target:
                continue

            for outer_right in range(nums_len - 1, outer_left + 2, -1):
                if (
                    outer_right < nums_len - 1
                    and nums[outer_right] == nums[outer_right + 1]
                ):
                    continue

                inner_left = outer_left + 1
                inner_right = outer_right - 1

                outer_sum = nums[outer_left] + nums[outer_right]

                while inner_left < inner_right:
                    curr_sum = outer_sum + nums[inner_left] + nums[inner_right]

                    if curr_sum == target:
                        four_sums.append(
                            (
                                nums[outer_left],
                                nums[inner_left],
                                nums[inner_right],
                                nums[outer_right],
                            )
                        )

                        while (
                            inner_left < inner_right
                            and nums[inner_left] == nums[inner_left + 1]
                        ):
                            inner_left += 1
                        while (
                            inner_left < inner_right
                            and nums[inner_right] == nums[inner_right - 1]
                        ):
                            inner_right -= 1

                        inner_left += 1
                        inner_right -= 1
                    elif curr_sum < target:
                        inner_left += 1
                    else:
                        inner_right -= 1

        return four_sums
